### PR TITLE
[utils] base_url: URL paths can contain &.  Backport from yt-dlp fixes #31485

### LINKED ISSUE #31485: Resources parsed from MPD files dont contain signature for request resulting in a HTTP 403 (DiscoveryPlus)
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2022.11.11** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

```
➜  2021.12.17 git:(stable) ./bin/youtube-dl -Uv --no-geo-bypass --cookies discoveryplus.com_cookies.txt https://www.discoveryplus.com/video/homestead-rescue/hardcore-homesteading
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-Uv', '--no-geo-bypass', '--cookies', 'discoveryplus.com_cookies.txt', 'https://www.discoveryplus.com/video/homestead-rescue/hardcore-homesteading']
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, pref UTF-8
[debug] youtube-dl version 2021.12.17
[debug] Git HEAD: 3748bed37
[debug] Proxy map: {}
It looks like you installed youtube-dl with a package manager, pip, setup.py or a tarball. Please use that to update.
yoooo im here using common downloader
[DiscoveryPlus] homestead-rescue/hardcore-homesteading: Downloading JSON metadata
[DiscoveryPlus] 846793: Downloading JSON metadata
hello there: ext: mpd format_id: dash
[DiscoveryPlus] homestead-rescue/hardcore-homesteading: Downloading MPD manifest
urlh.geturl:https://dplus-northamerica.media-edge.prod-vod.h264.io/82a6a9c8-8581-42f8-b83d-f7aec5daccf5/dash_clear_fmp4/edge-cache-token=Expires=9999999999&KeyName=media-edge-key-prod-northamerica-200d417&Signature=[REDACTED]/master.mpd
Traceback (most recent call last):
  File "/youtube-dl/2021.12.17/./bin/youtube-dl", line 33, in <module>
    sys.exit(load_entry_point('youtube-dl==2021.12.17', 'console_scripts', 'youtube-dl')())
  File "__init__.py", line 475, in main
    _real_main(argv)
  File "__init__.py", line 465, in _real_main
    retcode = ydl.download(all_urls)
  File "YoutubeDL.py", line 2070, in download
    res = self.extract_info(
  File "YoutubeDL.py", line 810, in extract_info
    return self.__extract_info(url, ie, download, extra_info, process)
  File "YoutubeDL.py", line 817, in wrapper
    return func(self, *args, **kwargs)
  File "YoutubeDL.py", line 838, in __extract_info
```
This bug is related to #4147 but that fix didnt quite resolve the issue.  The signature and expiration from the .mpd file is getting dropped when parsing the xml and appending the resources to the base_url.  I'm not sure what the proper fix is, but I was able to fix it with this kludge of a hack:

```python
    def _extract_mpd_formats(self, mpd_url, video_id, mpd_id=None, note=None, errnote=None, fatal=True, data=None, headers={}, query={}):

[...]
        mpd_base_url = base_url(urlh.geturl())


        mpd_base_url = urlh.geturl().replace("master.mpd", "")
```

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
➜  2021.12.17 git:(stable) ./bin/youtube-dl -Uv --no-geo-bypass --cookies discoveryplus.com_cookies.txt https://www.discoveryplus.com/video/homestead-rescue/hardcore-homesteading
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-Uv', '--no-geo-bypass', '--cookies', 'discoveryplus.com_cookies.txt', 'https://www.discoveryplus.com/video/homestead-rescue/hardcore-homesteading']
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, pref UTF-8
[debug] youtube-dl version 2021.12.17
[debug] Git HEAD: 3748bed37
[debug] Proxy map: {}
It looks like you installed youtube-dl with a package manager, pip, setup.py or a tarball. Please use that to update.
yoooo im here using common downloader
[DiscoveryPlus] homestead-rescue/hardcore-homesteading: Downloading JSON metadata
[DiscoveryPlus] 846793: Downloading JSON metadata
hello there: ext: mpd format_id: dash
[DiscoveryPlus] homestead-rescue/hardcore-homesteading: Downloading MPD manifest
urlh.geturl:https://dplus-northamerica.media-edge.prod-vod.h264.io/82a6a9c8-8581-42f8-b83d-f7aec5daccf5/dash_clear_fmp4/edge-cache-token=Expires=9999999999&KeyName=media-edge-key-prod-northamerica-200d417&Signature=[REDACTED]/master.mpd
Traceback (most recent call last):
  File "/youtube-dl/2021.12.17/./bin/youtube-dl", line 33, in <module>
    sys.exit(load_entry_point('youtube-dl==2021.12.17', 'console_scripts', 'youtube-dl')())
  File "__init__.py", line 475, in main
    _real_main(argv)
  File "__init__.py", line 465, in _real_main
    retcode = ydl.download(all_urls)
  File "YoutubeDL.py", line 2070, in download
    res = self.extract_info(
  File "YoutubeDL.py", line 810, in extract_info
    return self.__extract_info(url, ie, download, extra_info, process)
  File "YoutubeDL.py", line 817, in wrapper
    return func(self, *args, **kwargs)
  File "YoutubeDL.py", line 838, in __extract_info
```

### PR DESCRIPTION
**Pull request by Fuzion24**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [ ] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ ] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [ ] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [ ] Covered the code with tests (note that PRs without tests will be REJECTED)
- [ ] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [ ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

### CODE REVIEW

**Comment by dirkf** [comment]
Please fill in the description and include the magic line `Resolves #31485.`.

Anyway, why is `base_url()` using regex to parse URLs when there's an actual Python library module for that?

**Comment by dirkf** [comment]
Also, the change does have a test, so you can tick that checkbox too.

**Comment by Fuzion24** [comment]
I don't know. I didn't write any of this code. I'm just trying to get it to work. "Anyway, why is base_url() using regex to parse URLs when there's an actual Python library module for that?"

**Comment by dirkf** [comment]
Likewise. It was really a rhetorical question: the regex must have been an easy win.

Something like this makes it clearer what's being removed and what kept. It's not obvious what the right answer for (eg) `https://example.com/path1/path2` should be, vs. `https://example.com/path1/path2/`. The `url_basename()` function above gives `path2` for both, while the regex version of `base_url()` gives `https://example.com/path1/`, `https://example.com/path1/path2/`. A little specification comment would have been reassuring.
```py
def base_url(url):
    """
        Return the URL, with its path trimmed after the rightmost / 
        and without any query, params or fragment
    """
    parsed_url = compat_urlparse.urlparse(url)
    # url_basename() removes the last /
    # the regex pattern includes it?
    path = parsed_url.path.rsplit('/')[0]
    if not path.endswith('/'):
        path += '/'
    return compat_urlparse.urlunparse(parsed_url._replace(
        path=path, query='', params='', fragment='')
```

**Comment by dirkf** [comment]
Of course the "right answer" is what `test/test_utils.py` says:
```py
    def test_base_url(self):
        self.assertEqual(base_url('http://foo.de/'), 'http://foo.de/')
        self.assertEqual(base_url('http://foo.de/bar'), 'http://foo.de/')
        self.assertEqual(base_url('http://foo.de/bar/'), 'http://foo.de/bar/')
        self.assertEqual(base_url('http://foo.de/bar/baz'), 'http://foo.de/bar/')
        self.assertEqual(base_url('http://foo.de/bar/baz?x=z/x/c'), 'http://foo.de/bar/')
```

**Comment by dirkf** [comment]
And also, [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986) 2.2 categorises `&` as a reserved character of type `sub-delims`:
```
      sub-delims  = "!" / "$" / "&" / "'" / "(" / ")"
                  / "*" / "+" / "," / ";" / "="
```
The principle is that a `*sub*-delims` character like `&` need only be percent-encoded if it appears in a context where it has a delimiting role. Thus in `http://example.com/contains_an_&_here?param=contains_an_&_there`, the here `&` is fine but the there `&` must be rendered as `%26`, `http://example.com/contains_an_&_here?param=contains_an_%26_there`:
```
   URI producing applications should percent-encode data octets that
   correspond to characters in the reserved set unless these characters
   are specifically allowed by the URI scheme to represent data in that
   component.  If a reserved character is found in a URI component and
   no delimiting role is known for that character, then it must be
   interpreted as representing the data octet corresponding to that
   character's encoding in US-ASCII.
```
In other words, `base_url()` *should* support `&` in the path part of a URL.

**Comment by joaotolovikeepers** [comment]
Please approve this PR, the error is affecting many systems @dirkf
